File: src/java/infra/src/main/java/com/cdk/inventory/ordering/InventoryOrderingService.java
package com.cdk.inventory.ordering;

import com.cdk.constructs.InstrumentedFunction;
import com.cdk.constructs.InstrumentedFunctionProps;
import org.jetbrains.annotations.NotNull;
import software.amazon.awscdk.RemovalPolicy;
import software.amazon.awscdk.Tags;
import software.amazon.awscdk.services.lambda.IFunction;
import software.amazon.awscdk.services.lambda.eventsources.SnsEventSource;
import software.amazon.awscdk.services.lambda.eventsources.SnsEventSourceProps;
import software.amazon.awscdk.services.logs.LogGroup;
import software.amazon.awscdk.services.logs.LogGroupProps;
import software.amazon.awscdk.services.sqs.Queue;
import software.amazon.awscdk.services.sqs.QueueProps;
import software.amazon.awscdk.services.stepfunctions.*;
import software.constructs.Construct;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;

public class InventoryOrderingService extends Construct {
    public InventoryOrderingService(@NotNull Construct scope, @NotNull String id, @NotNull InventoryOrderingServiceProps props) {
        super(scope, id);

        LogGroup workflowLogGroup = new LogGroup(this, "JavaInventoryOrderingWorkflowLogGroup", LogGroupProps.builder()
                .logGroupName(String.format("/aws/vendedlogs/states/JavaInventoryOrderingWorkflowLogGroup-%s", props.sharedProps().env()))
                .removalPolicy(RemovalPolicy.DESTROY)
                .build());

        String workflowFilePath = "../infra/src/main/java/com/cdk/inventory/ordering/workflows/workflow.sample.asl.json";

        StateMachine workflow = new StateMachine(this, "JavaInventoryOrderingWorkflow", StateMachineProps.builder()
                .stateMachineName(String.format("JavaInventoryOrderingService-%s", props.sharedProps().env()))
                .definitionBody(DefinitionBody.fromFile(workflowFilePath))
                .logs(LogOptions.builder()
                        .destination(workflowLogGroup)
                        .includeExecutionData(true)
                        .level(LogLevel.ALL)
                        .build())
                .build());

        Tags.of(workflow).add("DD_ENHANCED_METRICS", "true");
        Tags.of(workflow).add("DD_TRACE_ENABLED", "true");

        HashMap<String, String> functionEnvVars = new HashMap<>(2);
        functionEnvVars.put("DD_SERVICE_MAPPING", String.format("lambda_sns:%s", props.newProductAddedTopic().getTopicName()));
        functionEnvVars.put("STEP_FUNCTIONS_WORKFLOW_ARN", workflow.getStateMachineArn());
        
        String compiledJarFilePath = "../inventory-ordering-service/target/com.inventory.ordering-0.0.1-SNAPSHOT-aws.jar";
        IFunction handleProductAddedFunction = new InstrumentedFunction(this, "ProductAddedFunction",
                new InstrumentedFunctionProps(props.sharedProps(), "com.inventory.ordering", compiledJarFilePath, "handleNewProductAdded", functionEnvVars)).getFunction();
        workflow.grantStartExecution(handleProductAddedFunction);
        
        handleProductAddedFunction.addEventSource(new SnsEventSource(props.newProductAddedTopic(), SnsEventSourceProps.builder()
                .deadLetterQueue(new Queue(this, "NewProductAddedEventSourceDLQ", QueueProps.builder()
                        .queueName(String.format("NewProductAddedEventSourceDLQ-%s", props.sharedProps().env()))
                        .build()))
                .build()));
        
    }
}

File: src/go/cdk/inventory-ordering-service/inventoryOrderingService.go
//
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.
//

package inventoryorderingservice

import (
	sharedprops "cdk/shared"
	sharedconstructs "cdk/sharedConstructs"
	"fmt"

	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awslambdaeventsources"
	"github.com/aws/aws-cdk-go/awscdk/v2/awslogs"
	"github.com/aws/aws-cdk-go/awscdk/v2/awss3assets"
	"github.com/aws/aws-cdk-go/awscdk/v2/awssns"
	"github.com/aws/aws-cdk-go/awscdk/v2/awssqs"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsssm"
	"github.com/aws/aws-cdk-go/awscdk/v2/awsstepfunctions"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type InventoryOrderingServiceProps struct {
	SharedProps       sharedprops.SharedProps
	ProductAddedTopic awssns.ITopic
}

func NewInventoryOrderingService(scope constructs.Construct, id string, props *InventoryOrderingServiceProps) {
	workflowLogGroup := awslogs.NewLogGroup(scope, jsii.String("InventoryOrderingWorkflowLogGroup"), &awslogs.LogGroupProps{
		LogGroupName:  jsii.Sprintf("/aws/vendedlogs/states/GoInventoryOrderingServiceLogGroup-%s", props.SharedProps.Env),
		RemovalPolicy: awscdk.RemovalPolicy_DESTROY,
	})

	workflow := awsstepfunctions.NewStateMachine(scope, jsii.Sprintf("GoInventoryOrderingWorkflow-%s", props.SharedProps.Env), &awsstepfunctions.StateMachineProps{
		StateMachineName: jsii.Sprintf("GoInventoryOrderingWorkflow-%s", props.SharedProps.Env),
		DefinitionBody:   awsstepfunctions.DefinitionBody_FromFile(jsii.String("./inventory-ordering-service/workflows/ordering-workflow.asl.json"), &awss3assets.AssetOptions{}),
		Logs: &awsstepfunctions.LogOptions{
			Destination:          workflowLogGroup,
			IncludeExecutionData: jsii.Bool(true),
			Level:                awsstepfunctions.LogLevel_ALL,
		},
	})
	awscdk.Tags_Of(workflow).Add(jsii.String("DD_ENHANCED_METRICS"), jsii.String("true"), &awscdk.TagProps{})
	awscdk.Tags_Of(workflow).Add(jsii.String("DD_TRACE_ENABLED"), jsii.String("true"), &awscdk.TagProps{})

	environmentVariables := make(map[string]*string)
	environmentVariables["ORDERING_SERVICE_WORKFLOW_ARN"] = jsii.String(*workflow.StateMachineArn())

	startWorkflowFunction := sharedconstructs.NewInstrumentedFunction(scope, "GoInventoryOrdering", &sharedconstructs.InstrumentedFunctionProps{
		SharedProps:          props.SharedProps,
		Entry:                "../src/inventory-ordering-service/handle-product-added/",
		FunctionName:         fmt.Sprintf("GoInventoryOrdering-%s", props.SharedProps.Env),
		EnvironmentVariables: environmentVariables,
	})
	startWorkflowFunctionDLQ := awssqs.NewQueue(scope, jsii.String("GoInventoryOrderingDLQ"), &awssqs.QueueProps{
		QueueName: jsii.Sprintf("GoInventoryOrderingDLQ-%s", props.SharedProps.Env),
	})
	startWorkflowFunction.Function.AddEventSource(awslambdaeventsources.NewSnsEventSource(props.ProductAddedTopic, &awslambdaeventsources.SnsEventSourceProps{
		DeadLetterQueue: startWorkflowFunctionDLQ,
	}))

	workflow.GrantStartExecution(startWorkflowFunction.Function)

	awsssm.NewStringParameter(scope, jsii.String("GoProductApiEndpoint"), &awsssm.StringParameterProps{
		ParameterName: jsii.String(fmt.Sprintf("/go/%s/inventory-ordering/state-machine-arn", props.SharedProps.Env)),
		StringValue:   jsii.String(*workflow.StateMachineArn()),
	})
}
